proofs/eo/cpc/rules/Rewrites.eo

```diff
@@ -1252,33 +1252,33 @@
   :args (s1 n1 m1 k1)
   :conclusion (= (>= k1 (seq.len (seq.extract s1 n1 m1))) true)
 )
-(declare-rule str-concat-clash ((s1 String) (s2 String :list) (t1 String) (t2 String :list))
-  :premises ((= (= s1 t1) false) (= (str.len s1) (str.len t1)))
+(declare-rule str-concat-clash ((@T0 Type) (@T1 Type) (@T2 Type) (@T3 Type) (s1 (Seq @T0)) (s2 (Seq @T1) :list) (t1 (Seq @T2)) (t2 (Seq @T3) :list))
```

Comment:
So seq.++ is defined in terms of str.++: 

`(define seq.++ () str.++)`

str.++ is declared here:

(declare-parameterized-const str.++ ((T Type))
  (-> (Seq T) (Seq T) (Seq T)) :right-assoc-nil ($mk_emptystr (Seq T)))
  
Showing that the arguments have to have the same type. The semantic of `:list` does not seem to change that. Then, the equality forces `($singleton_elim (seq.++ s1 s2))` and `($singleton_elim (seq.++ t1 t2)))` to have the same type.
  
So why is it necessary to specify 4 different types here? Is it for better matching or strictly necessary?

Reply:
It is not necessary to specify the 4 types here. However, the algorithm that auto-generates the Eunoia definition here is very basic and will not go out of its way to unify types.

We could in general improve the algorithm, but at the moment this was low priority since it just comes at the cost of a slightly more verbose Eunoia definition.